geowave/src/main/scala/geotrellis/spark/io/geowave/GeowaveLayerReader.scala

```diff
@@ -0,0 +1,243 @@
+      val tileSize = adapter.getTileSize
+      val tileLayout = TileLayout(maxCol - minCol + 1, maxRow - minRow + 1, tileSize, tileSize)
+      LayoutDefinition(extent, tileLayout)
+    }
+
+    val cellType = {
+      val geom = (new GeometryFactory).createPoint(new Coordinate((minX + maxX) / 2.0, (minY + maxY) / 2.0))
+      val queryOptions = new QueryOptions(adapter, index)
+      val query = new IndexOnlySpatialQuery(geom)
+      val config = computeConfiguration
+      GeoWaveInputFormat.setQuery(config, query)
+      GeoWaveInputFormat.setQueryOptions(config, queryOptions)
+
+      val gc = sc.newAPIHadoopRDD(
+        config,
+        classOf[GeoWaveInputFormat[GridCoverage2D]],
+        classOf[GeoWaveInputKey],
+        classOf[GridCoverage2D])
+        .map({ case (_, gc) => gc })
+        .collect.head
+
+      GridCoverage2DConverters.getCellType(gc)
+    }
+
+    val bounds = KeyBounds(
+      SpatialKey(0, 0),
+      SpatialKey(maxRow - minRow, maxCol - minCol) // sic
+    )
+
+    (TileLayerMetadata(cellType, layout, extent, LatLng, bounds), maxCol, minRow)
+  }
+
+  /**
+    * Read particular rasters out of the GeoWave database.  The
+    * particular rasters to read are given by the result running the
+    * provided LayerQuery.
+    *
+    * @param  id               The LayerId specifying the name and zoom level to query
+    * @param  rasterQuery      Produces a list of rasters to read
+    * @param  numPartitions    The number of Spark partitions to use
+    * @param  filterIndexOnly  ?
+    */
+  def read[
+    K: AvroRecordCodec: Boundable: JsonFormat: ClassTag,
+    V: AvroRecordCodec: ClassTag,
+    M: JsonFormat: GetComponent[?, Bounds[K]]
+  ](id: LayerId, rasterQuery: LayerQuery[K, M], numPartitions: Int, filterIndexOnly: Boolean) = {
+    import GeoWaveLayerReader._
+
+    if (!attributeStore.layerExists(id))
+      throw new LayerNotFoundError(id)
+    if (implicitly[ClassTag[K]].toString != "geotrellis.spark.SpatialKey")
+      throw new Exception("Unsupported Key Type")
+
+    /* XXX numPartitions, and filterIndexOnly are currently ignored. */
+
+    /* Boilerplate */
+    val LayerId(name, zoom) = id
+    val adapter = adapters.filter(_.getCoverageName == name).head
+    val strategy = substrats(zoom)
+    val ranges = strategy.getIndexStrategy.getHighestPrecisionIdRangePerDimension
+    val customIndex = new CustomIdIndex(strategy.getIndexStrategy, index.getIndexModel, index.getId)
+
+    /* GeoTrellis Metadata */
+    val (md, maxCol, minRow) = computeMetadata(adapter, ranges)
+
+    /* GeoWave Query and Query Options */
+    val queryOptions = new QueryOptions(adapter, customIndex)
+    val query = {
+      val fn = keyBoundsToGeometry(md.mapTransform, _: KeyBounds[SpatialKey])
+      val kbs = rasterQuery(md.asInstanceOf[M])
+      println(s"KEYBOUNDS=$kbs")
+      val geom = kbs
+        .map({ kb => fn(kb.asInstanceOf[KeyBounds[SpatialKey]]) })
+        .foldLeft(∅)({ (l, r) => l.union(r) })
```

Comment:
It is defined earlier in the [`GeoWaveLayerReader` companion object](https://github.com/geotrellis/geotrellis/pull/1542/files#diff-ab3b123f6ea96df4e71eb23b811595dbR42).